File: src/commands/giveaway.ts
import {
	ActionRowBuilder,
	ApplicationCommandType,
	ContextMenuCommandBuilder,
	MessageContextMenuCommandInteraction,
	ModalActionRowComponentBuilder,
	ModalBuilder,
	TextInputBuilder,
	TextInputStyle,
} from "discord.js";

export const data = new ContextMenuCommandBuilder()
	.setName("Giveaway")
	.setType(ApplicationCommandType.Message);

export async function execute(
	interaction: MessageContextMenuCommandInteraction
) {
	let peopleReacted: string[] = [];
	let winners: string[] = [];
	let replyMessage: string = "The giveaway winners are: ";

	const botID = interaction.client.application.id;

	const message = interaction.targetMessage;

	// Check if message has already processed before
	if (message.reactions.cache.get("787697278190223370")?.count! > 0) {
		await interaction.reply({
			content: "This message has been processed for giveaway",
			ephemeral: true,
		});
		return;
	}

	const modal = giveawayModal();
	await interaction.showModal(modal);

	interaction
		.awaitModalSubmit({ time: 30_000 })
		.then(async (modalInteraction) => {
			await modalInteraction.deferReply({
				ephemeral: true,
			});
			let numberOfWinners = parseInt(
				modalInteraction.fields.getTextInputValue("numberOfWinnersInput")
			);

			// Get the list of users who reacted
			const reactions = await message.reactions.cache;

			// People Who reacted towards the message
			for (const reaction of reactions) {
				const users = await reaction[1].users?.fetch();
				users.forEach((user) => {
					peopleReacted.push(user.id);
				});
			}

			// Remove the bot ID from the list of potential gardeners
			if (peopleReacted.indexOf(botID) > -1) {
				peopleReacted.splice(peopleReacted.indexOf(botID), 1);
			}

			// Remove duplicates from Array
			peopleReacted = peopleReacted.filter(
				(id, index) => peopleReacted.indexOf(id) === index
			);

			winners = peopleReacted.slice();
			winners.sort(() => Math.random() - 0.5);
			winners = winners.slice(0, numberOfWinners);

			for (let i = 0; i < numberOfWinners; i++) {
				replyMessage += `<@${winners[i]}> `;
			}

			await modalInteraction.editReply({
				content: `${replyMessage} - ${message.url}`,
				allowedMentions: { parse: ["users"] },
			});

			await message.react("787697278190223370");
		})
		.catch((err) => {
			console.error(err);
		});
}

function giveawayModal(): ModalBuilder {
	const modal = new ModalBuilder()
		.setCustomId("giveawayModal")
		.setTitle("Giveaway Modal");

	const numberOfWinnersInput = new TextInputBuilder()
		.setCustomId("numberOfWinnersInput")
		.setLabel("Number of people to win")
		.setStyle(TextInputStyle.Short)
		.setRequired(true);

	const firstActionRow =
		new ActionRowBuilder<ModalActionRowComponentBuilder>().addComponents(
			numberOfWinnersInput
		);

	modal.addComponents(firstActionRow);

	return modal;
}

File: lib/commands/giveaway.dart
import 'package:nyxx/nyxx.dart';
import 'package:nyxx_commands/nyxx_commands.dart';

final giveaway = MessageCommand(
  'Roll Giveaway winners',
  options: CommandOptions(defaultResponseLevel: ResponseLevel.hint),
  id(
    'Roll Giveaway winners',
    (MessageContext context) async {
      String replyMessage = 'The giveaway winners are: ';
      final clockeyId = context.interaction.applicationId;
      final message = context.targetMessage;

      await context.interaction.respondModal(_giveawayModal());
      final modalContext = await context.awaitModal(
        'giveawayModal',
        timeout: Duration(seconds: 30),
      );

      await modalContext.acknowledge();
      final numberOfWinners = int.parse(modalContext['numberOfWinnersInput']!);

      List<User> peopleReacted = [];
      final reactions = message.reactions;
      for (final reaction in reactions) {
        if (reaction.emoji is TextEmoji) {
          final users = await message.manager.fetchReactions(message.id,
              ReactionBuilder.fromEmoji(reaction.emoji as TextEmoji));
          peopleReacted.addAll(users);
        } else {
          final users = await message.manager.fetchReactions(
            message.id,
            ReactionBuilder(name: 'emoji', id: reaction.emoji.id),
            limit: 100,
          );
          peopleReacted.addAll(users);
        }
      }

      final ids = peopleReacted.map((user) => user.id.value).toList();

      // Remove the bots id from the potential gardener list
      if (ids.contains(clockeyId.value)) {
        ids.removeAt(ids.indexOf(clockeyId.value));
      }

      ids.shuffle();
      final winners = ids.toSet().take(numberOfWinners);

      for (final id in winners) {
        replyMessage += '<@$id> ';
      }

      await modalContext.respond(MessageBuilder(content: replyMessage));
    },
  ),
);

ModalBuilder _giveawayModal() {
  final numberOfWinnersInput = TextInputBuilder(
    customId: 'numberOfWinnersInput',
    style: TextInputStyle.short,
    label: 'Number of people to win the giveaway',
    isRequired: true,
  );

  final firstActionRow = ActionRowBuilder(components: [numberOfWinnersInput]);

  return ModalBuilder(
    customId: 'giveawayModal',
    title: 'Giveaway Modal',
    components: [firstActionRow],
  );
}
